better-code/src/chapter-2-contracts.md

```diff
@@ -0,0 +1,1044 @@
+```
+
+Predicates that work like `<=` are called *total preorders*.  In fact,
+we can convert any total preorder to a corresponding strict weak order
+(and vice-versa), with this function.
+
+```swift
+/// Returns the converse of `f`'s complement, `g(x, y) := !f(y, x)`.
+func converseOfComplement<T>(_ f: @escaping (T, T)->Bool)
+  -> (T, T)->Bool
+{
+  { x, y in !f(y, x) }
+}
+```
+
+Therefore, if we have a sorting implementation that works with any
+strict weak order, we can easily converti it to work with any total
+preorder by passing the predicate through `converseOfComplement`.
+
+
+Note that the name of the predicate became simpler: it no longer tests
+that its arguments represent an _increase_.  Instead, it tells us
+whether the order is correct.  Because the summary is no longer
+tricky, we can drop the example, and we're left with this:
+
+```swift
+/// Sorts the elements so that `areInOrder(self[i],
+/// self[i+1])` is true for each `i` in `0 ..< length - 2`.
+///
+/// - Precondition: `areInOrder` is a [total
+///   preorder](https://en.wikipedia.org/wiki/Weak_ordering#Total_preorders)
+///   over the elements of `self`.
+/// - Complexity: at most N log N comparisons, where N is the number
+///   of elements.
+mutating func sort<T>(areInOrder: (T, T)->Bool) { ... }
+```
+
+But we can go further and use a much simpler and more natural summary:
+
+```swift
+/// Sorts the elements so that all adjacent pairs satisfy
+/// `areInOrder`.
+```
+
+Usually, the less our documentation looks like code (without
+sacrificing precision), the greater its power as a tool for
+understanding programs, writing tests, and verifying
+correctness. [^argument-order].
+
+[^argument-order]: One could argue that to not lose precision, the
+    documentation needs to say in which order the adjacent elements
+    are passed to `areInOrder`. Judgement calls are part of
+    engineering, and in our judgement the order of arguments is
+    implied.
+
+There are two further simplifications we can make to the
+precondition. First, everything in documentation of a method happens
+in the context of `self`, so we can drop “of `self`.”  This
+improvement was available to us all along, and may seem trivial, but
+the value of [omitting needless
+words](https://thewritinghabit.blog/2016/04/04/omit-needless-words/)
+adds up over lots of function definitions. It lowers the burden of
+writing documentation, and makes it much easier to read. Remember,
+every word a reader has to process has a cognitive cost in addition to
+whatever benefits it may provide, so each one should more than pay for
+itself.
+
+Secondly, the summary has become so simple that we can embed the
+precondition there without overly complicating it, making the final
+declaration:
+
+```swift
+/// Sorts the elements so that all adjacent pairs satisfy the [total
+///   preorder](https://en.wikipedia.org/wiki/Weak_ordering#Total_preorders)
+/// `areInOrder`.
```

Comment:
To make this contract more approachable, could we add a brief, intuitive explanation of total preorders upfront? I'm concerned that requiring developers to click through to the Wikipedia article creates a high cognitive load for understanding the 90% use case.

Adding a simple analogy, like 'For context, a total preorder can be thought of as a generalization of arithmetic's less-than-or-equal-to (≤)', would make the contract easier to use correctly, saving time and helping to prevent bugs.